### DISCUSSION TITLE
In-source vs out-of-source use of cFS: Recommended project structure?

### DISCUSSION THREAD

**Discussion by stanislaw** [nasa/cFS#319] [Q&A]
Hello everyone,

We would like to know what a typical cFS-based project structure should look like. 

It looks like cFS directory structure assumes that the user apps should be created in the CFS_ROOT/apps directory which means that they should live inside a fork of the main cFS repository.

One alternative for us could be a vendor directory-based approach with the following structure:

Project root:

- apps
- sample_defs
- ... other stuff in our project
- vendor/
  - cFS/ -- unmodified cFS from the master branch
    - ...
    - cfe
    - osal
    - psp
    - ...
- CMakeLists.txt -- root CMake file of the project, not cFS (includes cFS via add_subdirectory)

On a previous project, we could achieve the vendor-based structure but it required serious CMake surgery on the normal cFS CMake files, and this was something we would like to avoid in the future.

The reasons why the out-of-source approach seems more attractive to us:

- Clean separation between the project-specific and the cFS repositories and file structures.
- No need to change anything inside the cFS folder. Also project-specific stuff is outside of the cFS source dir. Therefore no need to fork the cFS to create a project, also easier upgrades of the cFS if needed.

Is there a way we could achieve this out-of-cFS-source-tree configuration with the current cFS CMake structure? Also, does this approach seem reasonable?

Thank you for your attention.

**Comment by stanislaw** [comment]
This issue seems to be related: https://github.com/nasa/cFS/discussions/280.

**Reply by skliper** [reply]
The cFS repo was originally created to provide a sample set of submodules that work together and for CI (and an easy quickstart/example).  Really all you need is cfe/psp/osal within the context of your project (and your customized *_defs, whatever apps you want, etc).  There was actually a significant amount of debate related to setting up cFS as a monorepo vs submodules, which we really only got around by considering the cFS repo as an "example"... projects are free to start from it or create their own monorepo or whatever.  The intent wasn't for cFS to be taken as is and exist as a submodule within a project... which is why it ends up being a bit unfriendly to manage that way.

**Comment by mbenson1** [comment]
We have something similar.  We created our cmake build system before the NASA cmake build system was made public, and we still use ours.  We support many different projects, using different CFS versions and components, and need to have many different project repositories but be able to push and pull updates from the common repositories.  We also need to accommodate mixing public and proprietary code (apps, OSALs, PSPs, configurations, even forked CFE repos).  Our build system puts no constraints on where OSALs, PSPs, application code, tables, platform header files, or mission header files reside.  

We have a repository at https://github.com/windhoverlabs/baseliner, that creates our initial CFS project repository.  There is an input file where you specify the repository URLs and the commit/branch/tag you want to pull in for CFE, OSALs, PSPs, applications, and tools, and whether you want them as git modules or subtrees.  After you run the script, you push the newly created project to its new CM project.  We used that to create our https://github.com/windhoverlabs/airliner public repository, and repositories for other projects.  Some of the builds in that public repository require proprietary OSALs, PSPs, and applications.  Those are in private repositories, so those builds fail if one does not have access to the private repos.  The public builds still build though.

All components are relocatable, but our typical project structure is:

<main_private_repo>/public/   <- This is the Airliner repo at https://github.com/windhoverlabs/airliner
<main_private_repo>/public/apps   <- Sometimes we put apps in categorized subdirectories to organize them better.  Each app is either natively contained in the project repo, or a git module or subtree for shared applications.
<main_private_repo>/public/config  <- This can go arbitrarily deep.  Our build system allows for hierarchical configurations that inherit from the parent directory.
<main_private_repo>/public/config/ExampleA/Variant1  <- Example of nested configuration with some common configuration and some unique
<main_private_repo>/public/config/ExampleA/Variant2
<main_private_repo>/public/config/ExampleB/  <- Example of a totally different build configuration.
<main_private_repo>/public/config/ExampleB/sitl   <- Example of a SITL build that inherits the target configuration, but overrides whatever is required for a Software in the Loop configuration
<main_private_repo>/public/core/base/  <- Our fork of the CFE repo
<main_private_repo>/public/core/osal/  <- This contains project specific public OSALs, or shared OSALs as git modules or subtrees.
<main_private_repo>/public/core/osal/posix-nasa/  <- Our fork of the NASA POSIX OSAL.  
<main_private_repo>/public/core/osal/posix-wh/  <- Our POSIX OSAL.
<main_private_repo>/public/core/osal/... <- Some others
<main_private_repo>/public/core/psp/  <- This contains project specific public PSPs, or shared PSPs as either git modules or subtrees.
<main_private_repo>/public/core/psp/pc-linux-nasa  <- Our fork of the NASA PSP
<main_private_repo>/public/core/psp/pc-linux-wh  <- Our PC-Linux PSP
<main_private_repo>/public/core/psp/.... <- Some others
<main_private_repo>/public/core/tools/  <- Project independent tools that we always include.
<main_private_repo>/public/tools   <-  Project specific tools

<main_private_repo>/private/apps <- Proprietary or non-public applications
<main_private_repo>/private/config <- Proprietary or non-public header files, custom code, tables, cmake files, etc...
<main_private_repo>/private/core/osal/  <- Proprietary or non-public OSALs
<main_private_repo>/private/core/psp/  <- Proprietary or non-public PSPs
<main_private_repo>/private/tools  <- Proprietary or non-public project specific tools

<main_private_repo>/...  <- Non CFS stuff like drivers, OS, IP, etc...

My recommendation is to not be afraid to change the build system and be creative with where to put files to better fit your CM policies and work flow.  Most of the code is independent of the build system.  Version 6.7 did add some build system auto-generated code, but those are relatively easy to recreate.  

Lastly, I have also occasionally created totally different build structures to accommodate vendor specific build systems, i.e. Wind River Workbench, Green Hills Multi, Xilinx SDK, and Xilinx Vitis.  Green Hills works out well because build configuration is inheritable, making it easier to maintain, particularly for ARINC653 builds.  Workbench allows nested projects but configuration doesn't inherit so its difficult to maintain.  Xilinx SDK requires a lot of work.  Vitis was a total show stopper.

**Reply by jbohren-hbr** [reply]
@mbenson1 Thanks for pointing out your toolchain. I noticed that your toolchain sets `ELF2CFETBL_BIN` which is presumably used in your cFE fork instead of the way that the CMake toolchain in upstream cFE locates it ([as seen here](https://github.com/nasa/cFE/blob/d03e24e675fcd253222e40a5c544920e03854c5f/cmake/arch_build.cmake#L229)).

This is the kind of design feature that makes leveraging the existing CMake infrastructure difficult. The cFE source assumes a lot about where `elf2cfetbl` located, which breaks encapsulation in an unexpected way. This means that the cFS example repo organization is more than just an example. There are some other similar pitfalls, too, with which I'm sure the cFS team is familiar.

@skliper Is there any appetite for relaxing some of the assumptions in the main CMake toolchain about how the sources are organized? Even something as simple as adopting overrides such as one for where build tools like `elf2cfetbl` are located could be a step in the right direction.

**Reply by stanislaw** [reply]
I can confirm that I keep supporting two paths to `elf2cfetbl` when I build from the command-line vs build inside IDE when I want to run and debug the core-cpu1 binary without doing the install step. Having an override would be useful.

**Reply by jbohren-hbr** [reply]
It's actually surprising that `elf2cfetbl` isn't just in the cFE repo considering how tightly coupled it is with the cFE build process. It looks like it's the _only_ thing that forces a user into having a `${MISSION_SOURCE_DIR}/tools` directory in which it assumes it will find the source for `elf2cfetbl`.

**Reply by jphickey** [reply]
Worth noting that there is no specific requirement to use `elf2cfetbl` ... users are free to implement their own tool to generate table files.  The only "coupling" here is due to the (unfortunate) design decision that traditional CFE table contents are defined via C header files, and as such the conversion tool needs access to these header files in order to build the tool itself as well as to compile the table source with a C compiler (the first half of the two-part table build).  The actual invocation of `elf2cfetbl` is just to extract that single binary object from the ELF file, it isn't really tied to the build, except that it needs to use the same cross compiler, and use the same include path as from FSW.

I've long been advocating for a better way to define the "external" data structures consumed or produced by cFE - using C headers are just loaded with interoperability problems.    For example, in the EDS-enabled CFE build, the EDS also defines the table structure, and there is a tool called `eds2cfetbl` that replaces `elf2cfetbl`.  And because this tool does not need access to the FSW C headers (only EDS), table builds can be decoupled from the project build entirely, if you so choose.

**Reply by jbohren-hbr** [reply]
> @jphickey Worth noting that there is no specific requirement to use `elf2cfetbl` ... users are free to implement their own tool to generate table files.

Yeah, but without modifying the cFE cmake infrastructure, it still needs to be in a very specific location. The exact coupling I'm referring to is the explicit dependency on `elf2cfetbl` in [mission_build.cmake](https://github.com/nasa/cFE/blob/5e41330979586bcdb113f491ea6c38ff9975fc72/cmake/mission_build.cmake#L387-L391):
```cmake
  add_subdirectory(${MISSION_SOURCE_DIR}/tools tools)

  # Add a dependency on the table generator tool as this is required for table builds
  # The "elf2cfetbl" target should have been added by the "tools" above
  add_dependencies(mission-prebuild elf2cfetbl)
```

> @jphickey  I've long been advocating for a better way to define the "external" data structures consumed or produced by cFE - using C headers are just loaded with interoperability problems. For example, in the EDS-enabled CFE build, the EDS also defines the table structure, and there is a tool called `eds2cfetbl` that replaces `elf2cfetbl`.

Indeed. Being able to leverage all the work that's gone into the [EDS tooling](https://github.com/nasa/EdsLib) including `eds2cfetbl` without having to maintain our own fork of cFE is what's leading us down this path.

---

Edit: Changed source link to permalink

**Reply by skliper** [reply]
@jbohren-hbr - I'm in favor of relaxing assumptions where possible, especially if we can reduce some of the "magic" that is baked into the current setup.

Side note - there are rumors of a possible generic EDS related tool in work that could import/export EDS as well as associated binary files, support editing, visualization, etc.  The idea is it would support tables (generation/loads and dumps) as well as the binary "cFS files" that are produced by cFS as long as there is an associated data sheet, and hopefully be able to convert between the representations along w/ a simple editor/visualizer.  Other folks are working this so I may be out of date on the details or status (and the concepts may have evolved since my request/involvement).

**Reply by astrogeco** [reply]
Would creating a new cmake parameter, say `TABLE_TOOL_SOURCE_DIR`, be an acceptable, interim "solution"?

**Reply by jbohren-hbr** [reply]
> @astrogeco Would creating a new cmake parameter, say `TABLE_TOOL_SOURCE_DIR`, be an acceptable, interim "solution"?

Yeah, that's an approach that would relax the requirements on workspace layout in the near term.

From a dependency / scope standpoint, the `cfe_add_tables()` function really _wants_ to be defined along with the table tool sources themselves. Then, after finding the table source directory, the cFE CMake infrastructure could import implementations of those cmake functions which are tool-specific. Then cFE and table generation can be decoupled via the `cfe_add_tables()` interface. In that way, different implementations for `cfe_add_tables()` can be used depending on the mission config.

**Reply by astrogeco** [reply]
I opened nasa/cFE#1907 and one of our community members tried one approach to addressing it with nasa/cFE#1910. We're still on the fence whether to do a halfway solution or wait until cFS-Draco to implement something more fully fleshed out. Regardless, please check those out and add your thoughts.
